[FZEditor] Grid layout resizers fixes

SeraphimaZ opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
* Fixed
  * missing grid resizers
  * moving grid resizers after merge
  * gap between zones and taskbar

* Added possibility to move grid resizers along whole available space of row/column. 
* Refactored code and changed some data structures (raw arrays to lists)

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #2507, #2483, #2326, https://github.com/microsoft/PowerToys/issues/766, https://github.com/microsoft/PowerToys/issues/2440
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
* Missing grid resizers fix:
![2507_1 (Medium)](https://user-images.githubusercontent.com/8949536/83850972-55e0b700-a71a-11ea-8e80-f27608a0b096.png)

* Gap between zones was caused by calculation error after splitting zones. Zones are represented as a percentage of the whole screen, and after calculations total percentage could be less than original `10000`. 
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8949536/83857006-09e64000-a723-11ea-8d90-5a9c0dcf2b5c.gif)

* After merge zones redundant data wasn't removed. This caused the problem that users can't move dividers over the other dividers that are still present in data structures, but invisible for users.
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8949536/83858052-8decf780-a724-11ea-8ede-4faa09b2dd31.gif)

* Dividers now can be moved through whole available area
![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/8949536/83858546-56cb1600-a725-11ea-8b12-29927e7addc6.gif)

ivan100sic commented:
There are some issues:

+ After merging zones, zone numbers are chaotic.
+ After creating a grid layout with 4 zones and merging zones 2, 4, the editor crashes.

Another thing that could be a part of this PR is that the Merge Zones button should disappear if it's not used, e.g. when it loses focus

SeraphimaZ commented:
> There are some issues:
> 
> * After merging zones, zone numbers are chaotic.
> * After creating a grid layout with 4 zones and merging zones 2, 4, the editor crashes.
> 
> Another thing that could be a part of this PR is that the Merge Zones button should disappear if it's not used, e.g. when it loses focus

Thanks, I'll look into it

ivan100sic commented:
Now merging 2 and 4 in a 4-grid doesn't crash but it doesn't produce the correct layout. Zone numbers seem to work. However I still ended up with a crash after a few steps, let me try to find exact repro steps.

ivan100sic commented:
This might help you somehow: All of the incorrect layouts were formedHappened after performing a merge of two zones one above the other.

SeraphimaZ commented:
@ivan100sic thanks for the tip for checking incorrect behavior, I've fixed that. PTAL

yuyoyuppe commented:
I've tested the editor and couldn't crash it. The functionality seems to work. I've found a minor issue with a trembling zone heights if you align to rows slowly. I've sent a video to @SeraphimaZ

ivan100sic commented:
Another bug:

Start with a 3 row layout, with "Show space around zones" set to true and space set to 16, then create three vertical splitters as shown here (the two red ones are aligned):
![image](https://user-images.githubusercontent.com/9358361/84894063-86244000-b0a0-11ea-875e-3a652adb7e9c.png)

Then, drag the topmost vertical splitter to the right, past the green one. then left past the other two - the handle on the second resizer (marked green) will get detached.

crutkas commented:
is this fix going to work in 0.19?

enricogior commented:
@crutkas 
moved to 0.20

ivan100sic commented:
Still there's an issue with [this scenario](https://github.com/microsoft/PowerToys/pull/4095#issuecomment-645325042).

This time, start pulling the top resizer to the left, it will get detached immediately.

SeraphimaZ commented:
I've made update for fixing [this issue](https://github.com/microsoft/PowerToys/pull/4095#issuecomment-645325042)
Since resizers are not adjacent they should move separately. I've added this possibility and fixed some related bugs.
@ivan100sic PTAL

![detach](https://user-images.githubusercontent.com/8949536/86937905-79978280-c148-11ea-880d-0d4e0939a879.gif)

ivan100sic commented:
![image](https://user-images.githubusercontent.com/9358361/87026105-8e712600-c1db-11ea-928e-994bb06a6b37.png)

+ Start with a three row layout
+ Create vertical splits in the order 1, 2, 3 (3 is aligned with 1)
+ Merge two zones marked gray

This results in a crash.

SeraphimaZ commented:
> * Start with a three row layout
> * Create vertical splits in the order 1, 2, 3 (3 is aligned with 1)
> * Merge two zones marked gray
> 
> This results in a crash.

Thanks for the catch, fixed

ivan100sic commented:
+ Open the editor
+ Set the number of zones to 4
+ Start editing the grid layout
+ Merge the two zones on the left
+ Click cancel
+ Increment the number of zones by clicking +

The editor crashes.

ivan100sic commented:
+ Turn on Show space around zones and set the space to 16
+ Start editing the grid layout with 9 zones
+ Create two aligned vertical splitters as shown
+ Pull one of them to the left

The editor crashes.

![image](https://user-images.githubusercontent.com/9358361/87294490-2b95cc80-c504-11ea-90f5-81089acb20bc.png)

ivan100sic commented on src/modules/fancyzones/editor/FancyZonesEditor/GridDragHandles.cs:
```diff
@@ -0,0 +1,604 @@
+            }
+        }
+
+        private bool UpdateDragHanlderForExistingSplit(Orientation orientation, Func<GridResizer, bool> cmpr, Func<GridResizer, bool> endCmpr, Func<GridResizer, bool> startCmpr)
```
Typo in function name `Hanlder`

SeraphimaZ replied:
fixed

enricogior commented:
Great work, let's merge it!
